### ISSUE TITLE
[question] update cucumber to 2.x -> reports empty

### ISSUE THREAD

**Issue by toepi** [cukespace/cukespace#118]
I've try to update cucumber, currently to 2.4.0. But now I'm at a point where my understanding of the code ends:
In `CucumberReporter` around line 108:
`FeatureParser.parse(jsonReport)`
I get null?
I think it is the same issue as my main problem, which is that all json files under `target/cucumber-report` are empty.

Can you give me a hint how this reports are generated?

In `BaseCukeSpace` I must replace `new JsonFormater` with `(Formatter) new PluginFactory().create("json")` - should be the same only the creation is different?

**Comment by rmannibucau** [comment]
Hello @toepi , last cucumber is 4.2, i will have a look very soon if we can upgrade but API has been broken so it needs some more integration

**Comment by toepi** [comment]
Thats right but in 2.x they have make massive restructuring, so I hope small steps.

After upgrade we can replace your custom event api with eventbus (cucmberruntime.getEventBus())

**Comment by rmannibucau** [comment]
@toepi hmm i don't think so, here we don't trigger cucumber events but arquillian event which didnt change

**Comment by toepi** [comment]
right, I mean not this api, I mean the implementation inside the cucumberruntime, were you override the hooks and so one ... most of this parts are removed in cucumber 2 ... or replace with eventbus?

**Comment by rmannibucau** [comment]
@toepi yep, exact same applies to cucumber 4 so doing 1 -> 2 -> 4 does likely not worth it and i'm more tempted to do 1 -> 4. Long story short the need of these hooks were coming from the need to trigger events for arquillian extension since cucumber breaks JUnit integrations (there is not "test" method). Since the lifecycle listener API changed we need to find the same hooks in the new event bus and wire it there.

After a quick check it seems cucumber broke most of the API cukespace was relying on so my gut feeling is that starting back from cucumber.api.junit.Cucumber and merge the new logic with BaseCukeSpace/ArquillianCucumber is probably the best option today. We have the testng abstraction to maintain but the same hack we already have should do the trick (with BaseCukeSpace).

The main issue is that cucumber-jvm is paranoiac in its scoping so we have to fork a lot of code (step definition, jsonformatter since we can't use the plugin factory cause we want to control the output, etc...) but it is a straight line.

Do you want to give it a try? 

FYI i did my test on <cucumber-jvm.version>4.2.0</cucumber-jvm.version>

**Comment by toepi** [comment]
You mean this issue with the empty json-files is related to the new JsonFormater? But were on the current code the overriden lives? Or let me ask it other, were we get it's output?

Yes will try it. And yep I can do it also with 4.2!

But I've start with cucumber.Runtime? Bad start?

After your explainations we will never have a change to use Cucumbers JUnitRunner as delegate in CukeSpace (aka ArquillianCucumber)? That was the first try to have an easier migration, but after a short test, we must change all examples ...

**Comment by toepi** [comment]
oh, now I see were I remove some stuff needed to transfer the json content to the client .... try it tomorrow. Hope cucumber has not change the JsonFormater in a way that we can not get the result.

**Comment by rmannibucau** [comment]
It is no more public but we can hack it either at build time with asm or using the same package maybe

**Comment by toepi** [comment]
read my mind? same package was my idea too.

Stupid question, exists a way to send (Arquillian-)events from server to client? Background: currently the formate rare not more than event-listener, so create an own event listener convert events into some think transferable and on the client revert the conversation and use the default Formatters .... only a stupid idea ....

**Comment by rmannibucau** [comment]
@toepi warp does it but it is very complex for what it does. What is your goal? normally the protocol in place (which is sufficient for cukespace and easier to debug/work with) is to enrich the archives with what is needed (any config) at package time (arquillian/shrinwrap one) then on server execute the tests and write back locally what is needed. Finally the client can read the local state. it works with all "local" instances (which means all servers). We can have a socket later on but it requires a lot of configuration and would likely happen after this upgrade which will already require a lot of refactoring so surely saner to not break it all at once.

**Comment by toepi** [comment]
Take a look at the current reporter/formater. The all want a kind of an uri? But internally there are not more than event-listener. So instead of create a helper to create a formater on serverside let write it to a file or a stingbuilder -> catch events in own formater, put it back to client and on the client-side use cucumbers formater to handle the events? Think a less ugly hack. But yes not know if is so complicated.

**Comment by rmannibucau** [comment]
Hmm, means you fake 2 cucumbers runtimes at the end with 2 states managements + this network thing to handle portably. Since the reporter can output it anyway and most of the time requires some local state - even based on threadlocal for some custom ones - it sounds like something we would need to toggle off by default.

However i like the idea and we can surely reuse servlet transport (through a before filter+ few hacks to bypass security) and have a remote cucumber executer - which is exactly what arquillian is for test engines). That said it does not solve the issues and is still work on top of the upgrade since we need a formatter configured specifically for us and we need to get rid of the old lifecycle style anyway to upgrade IMHO.

So in short there is:

1. The upgrade
2. The enhancement to support a real client/server deployment for all our config

Doing 2 in 1 will make reviews and tests harder but is an option if you feel it is easier than writing a file locally as we already do as a first step

**Comment by toepi** [comment]
your are right again. first must be the upgrade. but after get an empty file now I have empty json `[]`, seem cucumber doesn't inform plugins if we use FeatureRunner :(.

Will do it with 2.4 for the first time, cucumber has remove more stuff we depends on. so we can start from beginning  ... and with 2.4 I have a compiling, working code expect reports. after fix this big issue I will get to the next major in hope the change not all again.

but need a lot of time, must read the cucumber code to see what they have change at here junitrunner to get the reports.

**Comment by toepi** [comment]
was a "racecondition" the runtimeoption check if is has a eventbus and if one available than this will be attached to the formatter. But the eventbus will be set by the runtime. After move report configuration a few line lower we have an invalid adoc now.

Another question why you have custom runtimeoptions? they create problems. after remove the code all tests run, with this code some tests crashed.

**Comment by rmannibucau** [comment]
Custom options? We reuse cucumber ones but enable some config in arquillian.xml and change defaults only IIRC

**Comment by toepi** [comment]
with custom runtimeoptions:
```
  CukesInBellyFrTest.performInternalCucumberOperations » NullPointer
  EsTest.performInternalCucumberOperations » NullPointer
  ScenarioOutlineTest.performInternalCucumberOperations » NullPointer
  J8Test.performInternalCucumberOperations » NullPointer
```
without all tests running without any issue. and why not stay closer to the original? if i remind me right they have also a way to run a seperate szenario or any failed szenario ... unsure for tags.
